### ISSUE TITLE
Refactored Stream Handling how do you want it?

### ISSUE THREAD

**Issue by jeichorn** [repejota/phpnats#131]
I need a version of the nats client that handled timeouts in a reasonable manner.

I'm still doing some additional testing, but I have a refactored version of the socket code in my fork, and an additional implementation using non blocking streams.  If your interested in I can try to push things into a reviewable set of pull requests.

https://github.com/pagely/phpnats

**Comment by sazo** [comment]
Interesting and a very good idea with the non-blocking for the sake of performance. Does your implementation support TLS?

**Comment by jeichorn** [comment]
It does but I ran into some stability problems at production scale that I couldn't solve without a larger rewrite.  Though the better timeout handling in this code might still be useful.  I needed to integrate things into a https://reactphp.org/ mainloop to meet all my goals, and the full rewrite to the reactphp socket lib was just too much.

I ended rewriting everything in golang to have a stable nats library when combined with a websocket lib.

https://github.com/pagely/phpnats/blob/develop/src/Nats/SocketNonBlocking.php
